### ISSUE TITLE
Diagnose extremely slow generate_hash_key time

### ISSUE THREAD

**Issue by kwaegel** [mozilla/sccache#497]
I have an (unfortunately private) project that is seeing very little speedup from `sccache`. Checking the server logs, `generate_hash_key` is taking an extremely long time, upwards of 30-200 seconds per file. Is there a good way to diagnose what is going on here?

My initial guess is that the preprocessor is just extremely slow, so the lack of direct mode is hurting me (`ccache` builds the entire project in 3 minutes with a hot cache), but I'm not sure how to verify this.

**Comment by apriori** [comment]
@kwaegel How did you diagnose this? I'd be curious to check hash creation times for my own project. I see questionable slow speed ups as well. I work on a multi-platform project. CCache builds improve build times from 5 minutes to 20 seconds, however, I see nothing similar so far with sccache in windows (did not check against sccache linux, yet).

**Comment by apriori** [comment]
Forget the remark, I could diagnose it as well now (using RUST_LOG="sccache::compiler::compiler=debug").

Do you happen to have code with very deep header include hiearchies? Or in other words: Is an expanded (with the preprocessor) translation unit super large for your project?

**Comment by apriori** [comment]
See the just created connected issue #504.

**Comment by apriori** [comment]
@kwaegel 
This will take a bit (got a functional local version with wyhash, but pretty much zero speed up). Something different seems to be going on here. Btw, what is your platform/os? Windows 64 bit as well?

**Comment by apriori** [comment]
@kwaegel 
Set RUST_LOG="sccache::compiler::c=trace, sccache::compiler::compiler=debug" and investigate these lines

```
ccache::compiler::c: [bla.cpp.obj]: Preprocessor output is 4598766 bytes
sccache::compiler::compiler: [bla.cpp.obj]: generate_hash_key took 1.494 s
```
What about is your size/time ratio?

**Comment by apriori** [comment]
Ok, this is weird output:

```
[0m 2019-08-27T19:09:39Z: sccache::compiler::c: [bla.cpp.obj]: Preprocessor output is 5040185 bytes
[0m 2019-08-27T19:09:39Z: sccache::util: Hashed 0 files in 2.836 s
[0m 2019-08-27T19:09:39Z: sccache::compiler::compiler: [bla.cpp.obj]: generate_hash_key took 2.837 s
[0m 2019-08-27T19:09:39Z: sccache::compiler::compiler: [bla.cpp.obj]: Hash key: 204970d2a9f06737
[0m 2019-08-27T19:09:39Z: sccache::compiler::compiler: [bla.cpp.obj]: Cache hit in 0.038 s
```

**Comment by apriori** [comment]
Yes it is. Measuring futures invocation without a composable abstraction (hello monad stacks?) is a pain as it becomes very intrusive and therefore error prone.
However, it seems we were just mislead by these logs. A slight extension to the logs lead to this

```
[0m 2019-08-27T20:09:56Z: sccache::util: Command invocation took 7.121 s
[0m 2019-08-27T20:09:56Z: sccache::compiler::msvc: Weird processing took 0.000 s
[0m 2019-08-27T20:09:56Z: sccache::util: Hashed 0 files in 7.121 s
[0m 2019-08-27T20:09:56Z: sccache::compiler::c: actual generate_hash_key took 0.001 s
[0m 2019-08-27T20:09:56Z: sccache::compiler::compiler: [bla.cpp.obj]: generate_hash_key took 7.124 s
[0m 2019-08-27T20:09:56Z: sccache::compiler::compiler: [bla.cpp.obj]: Cache hit in 0.040 s
```
So the runtime was completely dominated by running the preprocessor. Even more funny:
When attempting to time the exact commandline used for the respective file, I end up (with pipe to nul) with ridiculous 18.91s. I guess this is some additional weirdness of the windows console host. However: If I instead preprocess to a file, that takes just 3.8 seconds. So I wonder, might it be a viable approach to preprocess to a file and read that in windows?

**Comment by apriori** [comment]
Additional remark: Even running MSVC with "/Zs" (syntax only) combined with "/showincludes" results in nearly identical runtime as full preprocessing. So unless one would like to roll a completely separate parser, this would be the lower bound for the execution time on windows.

For my case I get (due to usage of BOOST and Eigen) 4784 includes for a single translation unit (of which 2337 are unique). So even if one would implement something like direct mode for sccache, these duplications would need to be cached as well so one does not calculate insane amounts of redundant hashes (context https://github.com/mozilla/sccache/issues/219)

**Comment by apriori** [comment]
Yes, the sad thing is that this is merely a 200 LOC codebase. And all the file in question does is using `boost.filesystem` and `Eigen/Dense` and it immediately blows up in your face.

**Comment by apriori** [comment]
@kwaegel What is your ccache performance if you force it to not use direct mode?
